### ISSUE TITLE
'Segmentation fault' when using ct.Kinetics load yaml file

### ISSUE THREAD

**Issue by 12Chao** [Cantera/cantera#844]
**System information**

- Cantera version: 2.5.0a2
- OS: Ubuntu 18.0
- Python version: 3.8.0

**Expected behavior**
Using command `ct.Kinetics (infile='cantera.yaml')`  to load yaml file aimed to calculate rate coefficients later. The expected behavior is file loaded without error or maybe an error thrown because  lack of a thermo model.

**Actual behavior**
`Segmentation fault(core dumped)` error happened, and the python interpreter shut down by itself.

**To Reproduce**

1. Open 'python interpreter using `python` command
2. Run 
`import cantera as ct`
`ct.Kinetics(infile='cantera.yaml')`
3. See error
`Segmentation fault (core dumped)`

**Attachments**

Here is the yaml file in text format.
[cantera.txt](https://github.com/Cantera/cantera/files/4467255/cantera.txt)

**Comment by ischoegl** [comment]
I can reproduce this on`2.5.0a4`, but get no issues if importing the mechanism using
```
gas = ct.Solution('cantera.yaml')
```
The thermo data show discontinuities, but that's about it. I do agree that a segfaulting instantiation of `ct.Kinetics` is problematic. At the same time, I have questions about using `ct.Kinetics` by itself as it does require at least species names which are part of thermo specifications (is it more than just a vehicle to enable diamond inheritance?).

Interestingly, the same behavior is observed for
```
kin = ct.Kinetics('gri30.yaml')
```
and is not specific to `yaml`, as it happens for the previous inputs `gri30.cti` and `gri30.xml` as well.

**Comment by bryanwweber** [comment]
Yes, we would expect that thermo data is required for a `Kinetics` instance. The problematic behavior is more the segfault than not instantiating, as @12Chao noted in the OP:

> an error thrown because lack of a thermo model.

**Comment by 12Chao** [comment]
To follow up, it works fine when I use `ct.Solution('cantera.yaml')` to load it and call the kinetics instances(`reverse_rate_constants`, `forward_rate_constants`,`equilibrium_constants` etc.).

**Comment by ischoegl** [comment]
@bryanwweber ... that’s what I meant - a thermo model is implicit. My question really is that I can see someone trying to instantiate a `ThermoPhase`, but `Kinetics` or `Transport` by themselves do not make sense as they are instantiated in parallel to `ThermoPhase` and require a thermo phase instance to exist. Within the context of diamond inheritance for `Solution` it all comes together (it inherits from all three), but that doesn’t mean that `Kinetics` by itself is viable.

To make sure, I’m commenting on the inheritance structure as it is implemented right now, and how it may not necessarily reflect the intended usage of those objects. (Imho, for a clean implementation of a viable `Kinetics` class, it needs to inherit from `ThermoPhase`, and not from `SolutionBase`. If the current implementation is maintained, an error should be thrown if `Kinetics` is instantiated by itself.) The segfault is certainly a bug, but I’d like to clarify the actual intended use first. My own reading is that `Kinetics` and `Transport` are not viable.

**PS:** I have the sneaking suspicion that my work on the C++ Solution base may have an impact here. The `Kinetics` object can be made viable by ensuring the existence of a `ThermoPhase`. This check is likely missing at this point, and there doesn’t appear to be a unit test for the Python  `Kinetics` call. Again, I have serious questions about using the constructor by itself, as it does not reflect inheritance.

**Comment by bryanwweber** [comment]
@ischoegl I think we are agreeing, that there should be an error if `Kinetics` is instantiated without a corresponding `ThermoPhase` instance, rather than a segfault. Fixing the segfault should be the priority here, IMO.

I think whether it would be possible for `Kinetics` to stand on its own, without `Solution`, is a separate question. Another (hypothetical) option is to pass an instance of `ThermoPhase` to the `Kinetics` constructor, rather than inheritance. But whether having a separate `Kinetics` instance is useful at all should be answered first.

**Comment by ischoegl** [comment]
@bryanwweber .... 👍 that’s exactly what my question is here. The way things are implemented, throwing an error would supposedly preempt the segfault. My concern is deprecation as the behavior in question was likely legitimate in 2.4.

**Comment by bryanwweber** [comment]
@ischoegl I just checked 2.4.0 installed via conda and that also segfaults, so it was definitely not legitimate at that time. If you want to discuss changing the behavior of `Kinetics` so that it can be instantiated in a way other than by `Solution`, I think that should happen over at Cantera/enhancements.

**Comment by ischoegl** [comment]
@bryanwweber Yup, I just confirmed the same, so I'm somewhat relieved that I didn't cause a regression. Beyond, I do not see a requirement for a viable `Kinetics` object, so I'd suggest to just raise an error to prevent the segfault and update the documentation to clarify that it is not meant to be used by itself. The only potential use case I see would be for mix-and-match submechanisms, where the old approach is currently broken (see #821); it would still be covered by `Solution` though.

**Comment by ischoegl** [comment]
Finally, as I suspected, the following segfaults also (on 2.4):
```
ct.Transport('gri30.cti') 
```
Same rationale applies ...

**Comment by ischoegl** [comment]
Fix was simple, see #845

~On further thought, moving a list of species and associated methods to the C++ Solution object would be a simple (and logically consistent) way to enable stand-alone objects. This will require more work, and may make sense in addressing #821. *I'm not sure it's worth the effort though as I simply cannot see enough use cases*~
